hivemind/client/optim/collaborative.py

```diff
@@ -0,0 +1,260 @@
+from __future__ import annotations
+import warnings
+from dataclasses import dataclass
+from threading import Thread, Lock, Event
+from typing import Optional
+
+import torch
+import numpy as np
+
+from hivemind.dht import DHT
+from hivemind.client.optim.base import DecentralizedOptimizerBase
+from hivemind.client.averaging.weighted import WeightedAverager
+from hivemind.utils import get_logger, get_dht_time, run_in_background, ValueWithExpiration
+from hivemind.client.optim.performance_ema import PerformanceEMA
+
+logger = get_logger(__name__)
+LRSchedulerBase = getattr(torch.optim.lr_scheduler, '_LRScheduler', None)
+
+#TODO make sure that the weighted averager will NOT advertize for averaging if it is out of sync!
+# warn if expiration is too infrequent (measure mean time between steps)
+#TODO do not forget to setup client mode!
+#TODO make sure we can recover from zero_grad!
+#TODO actually use scheduler normally on step!
+#TODO enforce minimum period on reporting progress
```

Comment:
Either resolve this before merging the PR or create corresponding issues

Reply:
done